### PR TITLE
Implement Restricted Kernel SDK using traits

### PR DESCRIPTION
**Pull request by jul-sh**
This interface is different from the one design doc ([go/oak-rust-sdk](http://goto.google.com/oak-rust-sdk)), but I feel pretty strongly that it is the right one. This PR exists as an alternative to #4578.

Fundamentally what we're hoping to with the [KeyOrigin](https://github.com/project-oak/oak/blob/db81a53843788cd8945d7a5a571e5a31cbd706f4/oak_restricted_kernel_sdk/src/dice.rs#L60) enum is to provide different implementations of the similar functionality. 

Rust provides a feature for that in its type system, we should use it. Doing is more idiomatic, increases safety (the correct type of signer being used can now be validated by the type system), and allows permits more versatile mocks for testing.

### CODE REVIEW

File: oak_restricted_kernel_sdk/Cargo.toml
Comment (tiziano88) [inline oak_restricted_kernel_sdk/Cargo.toml:10]:
```suggestion
mock_attestation = ["oak_stage0_dice", "oak_restricted_kernel_dice"]
```

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (tiziano88) [inline oak_restricted_kernel_sdk/src/dice.rs:28]: Not sure about this naming -- is the definition of an attester just anything that can return evidence?

Comment (jul-sh) [reply oak_restricted_kernel_sdk/src/dice.rs:28]: Copied the naming from the design doc, but agreed it isn't ideal. 

> is the definition of an attester just anything that can return evidence?

yes.

`Evidence` may be a more descriptive term. cc @ipetr0v

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (jul-sh) [reply oak_restricted_kernel_sdk/src/dice.rs:147]: I think this is fine. From the perspective of the SDK user, this is basically just a crypto oracle struct. The inner implementation may own a key, but it doens't change the fact that the interface it provides is just that of an oracle. 

Perhaps `Oracle` would be more adept than `Handle`. But I didn't want to also get into naming after already deviating from the design doc with the trait interface. :)

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (ipetr0v) [inline oak_restricted_kernel_sdk/src/dice.rs:203]: Attester cannot be Instance vs Group, because it doesn't depend on the Key Provisioning. So it should just be Attester

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (jul-sh) [reply oak_restricted_kernel_sdk/src/dice.rs:28]: Personally I'm okay with the name, but one could argue that it doesn't quite meet the definition outlined in the RFC 

> The Attester role is assigned to entities that create Evidence that is conveyed to a Verifier.

this struct just retrieves evidence, it doesn't do any attesting.

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (jul-sh) [reply oak_restricted_kernel_sdk/src/dice.rs:203]: Disagree. I think it's helpful to state that this attester retrieves the instance evidence, even if we never implement a `GroupAttester` (which would rely key provisioning. 

The other `instance` structs do not depend on key provisioning either.

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (ipetr0v) [reply oak_restricted_kernel_sdk/src/dice.rs:28]: > this struct just retrieves evidence, it doesn't do any attesting.

This is just the SDK interface, and the creation is hidden behind it (even though the `Attester` itself just retrieves it). So from the point of view of an app developer - it creates an Evidence

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (ipetr0v) [reply oak_restricted_kernel_sdk/src/dice.rs:203]: I think the existence of an InstanceAttester implies that there will be a GroupAttester

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (jul-sh) [reply oak_restricted_kernel_sdk/src/dice.rs:28]: > So from the point of view of an app developer - it creates an Evidence

Disagree, even from the point of view of the app developer, the evidence precedes this struct, it merely exposes it. 

An developer may create an `InstanceSigner` first. Yet the private key it uses for signing is implicitly linked to the public key in the `Evidence` struct created later.

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (jul-sh) [reply oak_restricted_kernel_sdk/src/dice.rs:203]: And maybe there never will be. Even so, imo it avoids confusion by making it clear which type of evidence is exposed.

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (tiziano88) [inline oak_restricted_kernel_sdk/src/dice.rs:19]: I think there are some missing words from the first sentence?
Also, break the TODO into its own line if possible (perhaps the formatter always joins them though, right?).

Comment (tiziano88) [inline oak_restricted_kernel_sdk/src/dice.rs:19]:
```suggestion
// Use aliasing the to conform to the naming defined in the Oak SDK design doc. TODO(#3841): Rename
```

**Review by tiziano88 — APPROVED** [review]
Looks reasonable, but I'll let @ipetr0v take a closer look.

File: oak_restricted_kernel_sdk/src/dice.rs
Comment (jul-sh) [reply oak_restricted_kernel_sdk/src/dice.rs]: addressed. :) The formatter only auto joins them some of them of the time (specifically if the preceding line is too long). Moved it to it's own line for now, and reworded the first sentence.

**Comment by tiziano88** [comment]
@ipetr0v could you prioritize reviewing this PR? I believe @jul-sh is waiting for this to be merged in order to reach out to client teams so they can adopt the SDK

**Review by ipetr0v — APPROVED** [review]

**Comment by jul-sh** [comment]
Thanks for your reviews @ipetr0v @tiziano88!

**Pull request merged**
